### ISSUE TITLE
Better cache APIs needed.

### ISSUE THREAD

**Issue by zephyrbot** [zephyrproject-rtos/zephyr#2647]
**_Reported by Chuck Jordan:_**

For Zephr cache support, currently there is only:    include/cache.h.

The API here is not sufficient.

First off, CPUs can have 
    * I-cache (with ops like disable, enable, invalidate all, invalidate by line, invalidate region)
    * d-cache (with ops like disable, enable
          * invalidate (all, by line, region)
          * flush w/I invalidate (all, by line, region)
          * flush-invalidate together

Further, there can be L1 cache, L2 cache, etc.

Device drivers, such as Ethernet MAC, might need to flush-by-line prior to transmitting a packet, or might need to invalidate-by-line when receiving a packet.
It is too expensive to do these ops over entire cache.
Its also too expensive to take an interrupt when the cache op is done.

If device drivers that need to do cache-ops are to be portable from target to target, a good API here would be appreciated.

Another use of cache ops is when its a HARVARD architecture, and interrupt vector table is subject to both I-cache and d-cache. User can write the vector table using data bus, but might need to fetch from it using instruction bus. So cache ops are needed to flush out changes, and invalidate the I-cache in that area.

Cache ops are also needed for multi-CPU applications -- Asymmetric Multi Processing -- when CPUs share memory subject to d-cache.

If an MMU is present, this may influence d-cache implementation -- with TAGS.

(Imported from Jira ZEP-1153)

**Comment by andrewboie** [comment]
Need to distill from the above (old) conversation what the gaps are in current code.

In current code we have:

```
__syscall void sys_cache_flush(void *addr, size_t size);
__syscall void sys_cache_invd(void *addr, size_t size);
size_t sys_cache_line_size_get(void);
```

> Oh. That reminds me. Related to this, but probably a different API, are memory fence operations. Some targets allow for out-of-order writing or so called write-queue.

We have in include/toolchain/gcc.h:
```
433:#define compiler_barrier() do { \
434-    __asm__ __volatile__ ("" ::: "memory"); \
435-} while (false)
```

> But another variant is ld.di and st.di. When the .di is added, it produces an opcode that cause the ops to completely by-pass the d-cache.
> So, for example, if you had an I/O peripheral in the part of the address map subject to d-cache, you could do memory-mapped-io in this way and bypass the cache -- even though the cache would normally TRY to do something with such loads and stores.
> So one way to do this is to create ACCESSOR macros for every data-size for loading and store in this way -- with cache bypassed.
> Device driver could would then use the accessors to talk to the peripheral and NOT use ASSIGNMENT statements in C code.

We already have the family of functions in include/sys/sys_io.h that drivers are supposed to use for MMIO access, so I don't think we need to do anything here. Whether the underlying implementation needs to do something special is very arch-dependent.

> Can x86 even do a cache invalidate without flushing? WBINVD certainly can't, and CLFLUSH doesn't look like it either. But this is necessary in order to interact properly with DMA controllers that may, for example, update a single memory location repeatedly to indicate where to go to read the data, etc.

Yes this is tricky -- we have separate APIs for flush and invd now, but whether one implicitly does the other varies

**Comment by carlocaione** [comment]
I was looking into this since I'm implementing cache operations for AArch64 but I was wondering what's the correct approach for the current two users of the APIs already in place (x86 and ARC) if I decide to add the remaining bits. Should I add some weak functions for the missing parts?

**Comment by eugenecohen** [comment]
Question: are we planning on adding a formal DMA API akin to the subsystem in linux?  I ask because from a device driver perspective drivers cannot inherently know if cache flushing/invalidation is needed for a given device/system.  So an abstraction such as DMA map/unmap or common buffer allocation can allow the driver source to remain agnostic to whether a system supports cache coherent dma.

**Comment by laurenmurphyx64** [comment]
@nashif I see there's 2 PRs for this already merged, plus https://github.com/zephyrproject-rtos/zephyr/pull/50988 by @carlocaione. Is this complete for 3.3?

**Comment by carlocaione** [comment]
> @nashif I see there's 2 PRs for this already merged, plus #50988 by @carlocaione. @carlocaione Is this complete for 3.3?

yeah, I feel like we can close this one down.
